src/typemap.c

```diff
@@ -1050,17 +1040,18 @@ jl_typemap_entry_t *jl_typemap_insert(union jl_typemap_t *cache, jl_value_t *par
     newrec->min_world = min_world;
     newrec->max_world = max_world;
     // compute the complexity of this type signature
-    newrec->va = jl_is_va_tuple(type);
+    newrec->va = jl_is_va_tuple((jl_datatype_t*)ttype);
     newrec->issimplesig = (tvars == jl_emptysvec); // a TypeVar environment needs an complex matching test
     newrec->isleafsig = newrec->issimplesig && !newrec->va; // entirely leaf types don't need to be sorted
     JL_GC_PUSH1(&newrec);
+    assert(jl_is_tuple_type(ttype));
     size_t i, l;
-    for (i = 0, l = jl_field_count(type); i < l && newrec->issimplesig; i++) {
-        jl_value_t *decl = jl_field_type(type, i);
+    for (i = 0, l = jl_field_count(ttype); i < l && newrec->issimplesig; i++) {
+        jl_value_t *decl = jl_field_type(ttype, i);
```

Comment:
It looks like you also added support in `sig_match_simple` to handle the case where `jl_is_type_type(decl->body) && jl_tparam0(decl->body) == decl->var` (although it has an odd way of testing for it), although the corresponding support in `sig_match_by_type_simple` doesn't seem to be present. If you want that case to fall through to getting tested by that loop, you should add it here. Otherwise, it'll fallback to calling `jl_subtype`.